src/rust/bin/wallet_tool.rs

```diff
@@ -0,0 +1,528 @@
+                .map(|t| t.starts_with("# - recovery_phrase=\""))
+                .unwrap_or(false)
+        })
+        .collect();
+    if phrase_line.len() != 1 || phrase_line[0].is_err() {
+        return Err(io::Error::from(ErrorKind::InvalidData));
+    }
+    let phrase = phrase_line[0]
+        .as_ref()
+        .unwrap()
+        .trim_start_matches("# - recovery_phrase=\"")
+        .trim_end_matches('"');
+
+    // This panic hook allows us to make a best effort to clear the screen (and then print
+    // another reminder about secrets in the export file) even if a panic occurs.
+
+    let moved_export_path = export_path.to_string(); // borrow checker workaround
+    let old_hook = panic::take_hook();
+    panic::set_hook(Box::new(move |panic_info| {
+        clear_and_show_cautions(&moved_export_path);
+
+        let s = panic_info.payload().downcast_ref::<&str>().unwrap_or(&"");
+        eprintln!("\nPanic: {}\n{:?}", s, Backtrace::new());
+    }));
+
+    let res = (|| -> io::Result<()> {
+        println!("The recovery phrase is:\n");
+
+        const WORDS_PER_LINE: usize = 3;
+
+        let words: Vec<_> = phrase.split(' ').collect();
+        let max_len = words.iter().map(|w| w.len()).max().unwrap_or(0);
+
+        for (i, word) in words.iter().enumerate() {
+            print!("{0:2}: {1:2$}", i + 1, word, max_len + 2);
+            if (i + 1) % WORDS_PER_LINE == 0 {
+                println!();
+            }
+        }
+        if words.len() % WORDS_PER_LINE != 0 {
+            println!();
+        }
+
+        println!(concat!(
+            "\nPlease write down this phrase on something durable that you will keep\n",
+            "in a secure location.\n",
+            "Press Enter when finished; then the phrase will disappear and you'll be\n",
+            "asked to re-enter a selection of words from it."
+        ));
+
+        let mut stdin = io::stdin();
+        let mut buf = String::new();
+        prompt(&mut stdin, &mut buf)?;
+
+        // The only reliable and portable way to make sure the recovery phrase
+        // is no longer displayed is to clear the whole terminal (including
+        // scrollback, if possible). The text is only printed if clearing fails.
+        try_to_clear(concat!(
+            "\n\n\n\n\n\n\n\n\n\n\n\n",
+            "Please adjust the terminal window so that you can't see the\n",
+            "recovery phrase above. After finishing the backup, clear the\n",
+            "terminal window"
+        ));
+
+        println!("\nNow we're going to confirm that you backed up the recovery phrase.");
+
+        let mut rng = thread_rng();
+        let mut unconfirmed: Vec<usize> = (0..words.len()).collect();
+        for _ in 0..min(3, words.len()) {
+            let index: usize = rng.gen_range(0..unconfirmed.len());
+            let n = unconfirmed[index];
+            unconfirmed[index] = unconfirmed[unconfirmed.len() - 1];
+            unconfirmed.pop().expect("should be nonempty");
```

Comment:
This is confusing me, why not just do...

```rust
let index: usize = rng.gen_range(0..unconfirmed.len());
let n = unconfirmed[index];
unconfirmed.remove(index);
```

Reply:
`remove` is O(length of vector); this is O(1). It's a [Durstenfeld shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm).